nslsii/__init__.py

```diff
@@ -160,20 +165,49 @@ def configure_base(user_ns, broker_name, *,
         from ophyd import setup_ophyd
         setup_ophyd()
 
-    if not ophyd_logging:
-        # Turn on error-level logging, particularly useful for knowing when
-        # pyepics callbacks fail.
+    if configure_logging:
         import logging
-        import ophyd.ophydobj
-        ch = logging.StreamHandler()
-        ch.setLevel(logging.ERROR)
-        ophyd.ophydobj.logger.addHandler(ch)
+        from logging.handlers import TimedRotatingFileHandler
+
+        log_file_handler = TimedRotatingFileHandler(
+            filename=bluesky_log_file_path,
+            when="W0",
+            backupCount=10
+        )
+        log_file_handler.setLevel("INFO")
+
+        log_file_format = (
+            "[%(levelname)1.1s %(asctime)s.%(msecs)03d %(name)s  %(module)s:%(lineno)d] %(message)s"
+        )
+
+        log_file_handler.setFormatter(
+            logging.Formatter(fmt=log_file_format)
+        )
+        logging.getLogger("bluesky").addHandler(log_file_handler)
+        logging.getLogger("caproto").addHandler(log_file_handler)
+        logging.getLogger("ophyd").addHandler(log_file_handler)
+        logging.getLogger("nslsii").addHandler(log_file_handler)
+        get_ipython().log.addHandler(log_file_handler)
+
+        # set the loggers to send DEBUG and higher log
+        # messages to their handlers
+        logging.getLogger("bluesky").setLevel("DEBUG")
+        logging.getLogger("caproto").setLevel("DEBUG")
+        logging.getLogger("ophyd").setLevel("DEBUG")
+        logging.getLogger("nslsii").setLevel("DEBUG")
+        get_ipython().log.setLevel("DEBUG")
 
     if ipython_exc_logging:
         # IPython logging must be enabled separately
         from nslsii.common.ipynb.logutils import log_exception
         get_ipython().set_custom_exc((BaseException,), log_exception)
 
+        get_ipython().magic("logstart -o -t bluesky_ipython.log append")
```

Comment:
For users, including us when we're seated at a beamline, the most convenience way to browse IPython history is to use the [`%history magic`](https://ipython.readthedocs.io/en/stable/interactive/magics.html#magic-history) which actually has a ton of functionality, including simple-to-use search, as in

```
%history -g fccd  # every command ever with 'fccd' in it
```

For users, that's probably handier than a log file.

I thought the role of this log file was an easy way to us to review what has been done when trying to piece together bug reports after the fact. The history stored for `%history` is in the sqlite database, so it's not so convenient to inspect over SSH. And it omits outputs.

On Friday's Pilot call I advertised this logging work, and the ensuing conversation touched on the possibility of creating one `logstart` file per IPython session. That seems easier to navigate than rotating logs at arbitrary points in time, potentially bisecting a series of commands that comprise one "thought". And we could implement it pretty easily, such as

```py
from datetime import datetime
get_ipython().magic(f"logstart -o -t {bluesky_log_file_path}/bluesky_ipython_{datetime.now().iso_format()}.log")
```

Reply:
I gave a little thought to having one ipython log file per session but I decided against it for two reasons, one of which is pretty much the same reason given above for avoiding bisecting one "thought": there can be _many_ sessions per "thought" and having at most two if the log file (if the file gets rotated mid-thought) makes them _easier_ to navigate. If you want ease of navigation and limited log files I don't think you can do better than just two logs, a 'current' one and an 'old' one.

The second reason I don't like one-file-per-session as described is there is no clean up of old files and adding that is more complicated than checking one file's size.

Reply:
OK. When reading through a file that concats multiple sessions we will have to be aware of sessions boundaries because of course all the variables will be reset. Perhaps the log-concatenation mechanism you propose could add a very visible marker?

I'm game to run with whatever you think will be best.

Reply:
I have committed a proposed solution.